graphql/http.go

```diff
@@ -17,19 +19,61 @@ func HTTPHandler(schema *Schema, middlewares ...MiddlewareFunc) http.Handler {
 	}
 }
 
-// HTTPHandlerWithErrorHandling works as HTTPHandler
-// but in addition provides passing errorHandler func
-// which will catch errors happened outside middleware
-func HTTPHandlerWithHooks(schema *Schema, errorHandler outsideMiddlewareErrorHandlerFunc, successfulResponseHook responseHook, middlewares ...MiddlewareFunc) http.Handler {
+// HTTPHandlerWithHooks works as HTTPHandler but in addition provides passing
+// errorHandler func which will catch errors happened outside middleware
+func HTTPHandlerWithHooks(schema *Schema, maxRequests int, minRequests int, waitTime time.Duration, errorHandler outsideMiddlewareErrorHandlerFunc, successfulResponseHook responseHook, middlewares ...MiddlewareFunc) http.Handler {
 	return &httpHandler{
-		schema:                 schema,
-		errorHandler:           errorHandler,
-		middlewares:            middlewares,
-		successfulResponseHook: successfulResponseHook,
+		schema:                  schema,
+		maxRequests:             maxRequests,
+		minRequests:             minRequests,
+		currentMaxRequestsLevel: maxRequests,
+		waitTime:                waitTime,
+		errorHandler:            errorHandler,
+		middlewares:             middlewares,
+		successfulResponseHook:  successfulResponseHook,
 	}
 }
 
+type endRequestState int
+
+const (
+	maxSimultaneousRequestsDefault int             = 15
+	minSimultaneousRequestsDefault int             = 2
+	waitTimeDefault                time.Duration   = 3 * time.Second
+	endRequestStateOK              endRequestState = 0
+	endRequestStateError           endRequestState = 1
+	endRequestStateTimedOut        endRequestState = 2
+)
+
+// activeRequest provides structure for request that processing by service
+// stores id, count of simultaneous requests on moment of starting request,
+// date of request start and predicted request duration
+type activeRequest struct {
+	id          uuid.UUID
+	bucketLevel int
+	startedAt   time.Time
+	predictedAt time.Time
+}
+
 type httpHandler struct {
+	// requestsBucket a list of all simultaneous requests
```

Comment:
It might be nice to have a separate API for keeping track of the requests instead of doing it on the handler. So a separate object which you initialised and pass to the handler. Putting all this state inside the handler is mixing concerns a bit.